### PR TITLE
Parallelise migration OKAPI-939

### PR DESCRIPTION
**Pull request by adamdickmeiss**
https://issues.folio.org/browse/OKAPI-939

Migration is parallel for all modules with dependencies satisfied (not just those with no requires).

Sorting always for Disable and enable in DepResolution.sortTenantModules (so that simulate shows the order).
But in reality that's not required as it's sorted anyway in jobRunPending later.

The sorting is completely reworked again (sortTenantModules). But at least the strategy is same for simulate phase and "real" migration phase.

I wished I could turn sortTenantModules and JobRunPending into one method as they are similar.

### CODE REVIEW

**Comment by adamdickmeiss** [comment]
> > Should this be configurable - whether to perform migration in parallel?
> 
> Yes. Otherwise some installations might get overloaded causing timeouts. We don't know if there are modules with migrations that fail when run in parallel (for example assuming exclusive database access). And the dependency declarations might be incorrect.
> 
> A switch like parallelMigration=true is sufficient. A more advanced solution might be a configuration option parallelMigrations=3 that limits the number. parallelMigrations=1 is sequential.

Such a switch is now implemented with a default value of 1 for now.

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 1]
```diff
@@ -1050,6 +1035,99 @@ private ModuleDescriptor findSystemInterface(Tenant tenant, String interfaceName
...
 
+  private boolean isExclusive(ModuleDescriptor md) {
+    return md.getSystemInterface("_tenantPermissions") != null
+        || md.getAuthRoutingEntry() != null;
+  }
+
+  private boolean depsOK(TenantModuleDescriptor tm, ModuleDescriptor md,
+      Collection<ModuleDescriptor> modules, Set<String> allProvided) {
+
+    if (tm.getAction().equals(Action.disable)) {
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:1046]: Can you replace `.equals` by `==` when comparing enums for readability, better type check, and avoiding NPE?

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 2]
```diff
@@ -1050,6 +1035,99 @@ private ModuleDescriptor findSystemInterface(Tenant tenant, String interfaceName
...
+      List<TenantModuleDescriptor> tml, Map<String, ModuleDescriptor> modsAvailable,
+      Set<String> allProvided, InstallJob job, AtomicInteger running, AtomicBoolean exclusive,
+      Promise<Void> promise) {
+
+    Iterator<TenantModuleDescriptor> iterator = tml.iterator();
+    while (iterator.hasNext() && running.get() < options.getMaxParallel() && !exclusive.get()) {
+      TenantModuleDescriptor tm = iterator.next();
+      ModuleDescriptor md = modsAvailable.get(tm.getId());
+      if ((tm.getStage().equals(TenantModuleDescriptor.Stage.pending)
+          || tm.getStage().equals(TenantModuleDescriptor.Stage.deploy))
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:1062]: == replacing .equals, consider importing TenantModuleDescriptor.Stage.

File: okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java [hunk 3]
```diff
@@ -1050,6 +1035,99 @@ private ModuleDescriptor findSystemInterface(Tenant tenant, String interfaceName
...
+          && depsOK(tm, md, getEnabledModules(t), allProvided)
+          && (!isExclusive(md) || running.get() == 0)) {
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/managers/TenantManager.java:1064]: This complex if with 5 expressions combined with `|| && && && ||` can be simplified by using guard clauses; this also reduces indentation of the body of the if:

```
      if (   tm.getStage() != TenantModuleDescriptor.Stage.pending
          && tm.getStage() != TenantModuleDescriptor.Stage.deploy ) {
        continue;
      }
      if (! depsOK(tm, md, getEnabledModules(t), allProvided)) {
        continue;
      }
      ...
```

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 4]
```diff
@@ -735,7 +665,70 @@ static void installMaxIterations(
...
+                .collect(Collectors.joining(", ")));
+        if (DepResolution.moduleDepProvided(modules, allProvided, md)) {
+          logger.info("yes: adding {}", md.getId());
+          iterator.remove();
+          iterator = tml.iterator();
+          result.add(tm);
+          modules.add(md);
+          String moduleFrom = tm.getFrom();
+          if (moduleFrom != null) {
+            logger.info("yes: removing from {}", moduleFrom);
```
Comment (julianladisch) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:721]: sortTenantModules should log on debug level only.

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

**Comment by adamdickmeiss** [comment]
> Looks good, thanks!

Oh really? I'll look at the comments and address them tomorrow. Thx.

File: okapi-core/src/main/java/org/folio/okapi/bean/ModuleDescriptor.java [hunk 5]
```diff
@@ -232,6 +233,23 @@ public void setOptional(InterfaceDescriptor[] optional) {
     return all;
   }
 
+  /**
+   * Get auth routing entry if such exist for module.
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/bean/ModuleDescriptor.java:237]: Typo: change "exist" to "exists". Same below.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 6]
```diff
@@ -128,7 +128,6 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
 
   /**
    * Check if md's required and optional interfaces are provided by set of modules.
-   * The interface version is not checked, only the interface name.
    * Only interfaces listed in allProvided are checked, this allows to exclude some optional
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:132]: "exclusion of" is a little more idiomatic

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 7]
```diff
@@ -137,54 +136,56 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+        return false;
       }
     }
     return true;
   }
 
-  static void topoSort(List<ModuleDescriptor> modules) {
-    List<ModuleDescriptor> result = new LinkedList<>();
+  /**
+   * Check if md's provided interfaces are required by any of modules.
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:164]: "any of the modules"

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 8]
```diff
@@ -137,54 +136,56 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+   * Only interfaces listed in allProvided are checked, this allows to exclude some optional
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:165]: "for the exclusion of"

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 9]
```diff
@@ -137,54 +136,56 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+   * interfaces from the check.
+   *
+   * @param modules     the modules that check against
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:168]: "The modules to check against." -- just add capital and period.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 10]
```diff
@@ -137,54 +136,56 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+   * @param md          module to check.
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:169]: "Module to check."

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java [hunk 11]
```diff
@@ -137,54 +136,56 @@ public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
...
+   * @return            true if modules do not use any of the provided interfaces; false otherwise
```
Comment (steveellis) [inline okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:170]: Probably capitalize sentence and add period.

File: okapi-core/src/main/raml/okapi.raml [hunk 12]
```diff
@@ -46,6 +46,66 @@ traits:
...
+          all problem(s) with module upgrade(s).
+        type: boolean
+        required: false
+        default: false
+      invoke:
+        description: |
+         Whether to invoke for tenant init/permissions/purge. Use
+          "true" to invoke, "false" to not invoke. Any other value
+          is a regular expression that is matched against the module ID.
+          If that pattern matches, invoke is performed (same as "true)".
```
Comment (steveellis) [inline okapi-core/src/main/raml/okapi.raml:78]: "true")

**Review by steveellis — COMMENTED** [review]
Just noticed a few typos.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 9 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:168]: I don't use capitals anywhere for parameters as first letter. So doing that now would be a mix of styles.

Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:168]: Same story with ending period. The last sentence do not end in period.

File: okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java (hunk 10 above)
Comment (adamdickmeiss) [reply okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java:169]: I'll remove the ending period instead.

**Pull request merged**